File: 3934-CouponCodeValidator/3934-CouponCodeValidator.py
# Last updated: 12/25/2025, 7:10:41 PM
class Solution:
    def validateCoupons(self, code: List[str], businessLine: List[str], isActive: List[bool]) -> List[str]:
        ret = {}
        n = len(code)

        valid = ["electronics", "grocery", "pharmacy", "restaurant"]
        for v in valid:
            ret[v] = []
        
        for i in range(n):
            c = code[i]
            b = businessLine[i]
            active = isActive[i]

            if not active or b not in valid or len(c) == 0: continue
            okay = True
            for char in c:
                is_valid = char in 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
                is_valid = is_valid or char in '0123456789'
                is_valid = is_valid or char == '_'

                if not is_valid:
                    okay = False
                    break

            if okay: ret[b].append(c)

        ans = []
        for b in valid:
            ans += sorted(ret[b])

        return ans

            

File: 3934-CouponCodeValidator/3934-CouponCodeValidator.cpp
// Last updated: 12/25/2025, 7:10:37 PM
class Solution {
public:
    bool isValid(string str) {
        for (char chr: str) {
            int a = chr;
            if (!((a >= (int)'a' && a <= (int)'z') || (a >= (int)'A' && a <= (int)'Z') || (a >= (int)'0' && a <= (int)'9') || a == 95)) {
                return false;
            }
        }
        return true;
    }

    vector<string> validateCoupons(vector<string>& code, vector<string>& businessLine, vector<bool>& isActive) {
        int n = code.size();
        vector<pair<string, string>> ret;
        unordered_set<string> valid_buisness = {"electronics", "grocery", "pharmacy", "restaurant"};
        for (int i=0; i<n; i++) {
            if (code[i].length() && isValid(code[i]) && valid_buisness.find(businessLine[i]) != valid_buisness.end() && isActive[i]) {
                ret.push_back({code[i], businessLine[i]});
            }
        }

        sort(ret.begin(), ret.end(), [](auto& a, auto& b) {
            return a.second < b.second || (a.second == b.second && a.first < b.first);
        });

        vector<string> res;
        for (auto& it: ret) res.push_back(it.first);

        return res;
    }
};